### PR TITLE
MM-12364 - Added capability to modify the type of a team to either public or private

### PR DESCRIPTION
**Pull request by waseem18**
#### Summary
This PR add the capability to modify the type of a team to either public or private.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/9670

#### Checklist
- [ ] Added or updated unit tests (required for all new features)

### CODE REVIEW

**Comment by waseem18** [comment]
@grundleborg Please take a look this whenever you're free. 

Will add unit test tomorrow.

File: model/team.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ const (
 	TEAM_EMAIL_MAX_LENGTH           = 128
 	TEAM_NAME_MAX_LENGTH            = 64
 	TEAM_NAME_MIN_LENGTH            = 2
+	TEAM_PRIVATE                    = "P"
```
Comment (saturninoabril) [inline model/team.go]: Can we move this to the top, next to "I" (team_invite) just to group them together?

File: cmd/mattermost/commands/team.go [hunk 2]
```diff
@@ -339,3 +353,43 @@ func archiveTeamCmdF(command *cobra.Command, args []string) error {
...
+	if public == private {
+		return errors.New("You must specify only one of --public or --private")
+	}
+
+	team := getTeamFromTeamArg(a, args[0])
+	if team == nil {
+		return errors.New("Unable to find team '" + args[0] + "'")
+	}
+
+	if !(team.Type == model.TEAM_OPEN || team.Type == model.TEAM_PRIVATE) {
```
Comment (saturninoabril) [inline cmd/mattermost/commands/team.go]: @grundleborg Just like to know about the "I" (or TEAM_INVITE) type. Will it remain as is and will not be possible to modify after team creation?

File: app/team.go [hunk 3]
```diff
@@ -1188,3 +1189,15 @@ func (a *App) RemoveTeamIcon(teamId string) *model.AppError {
...
+	team, err := a.UpdateTeam(oldTeam)
+	if err != nil {
+		return team, err
+	}
+	messageWs := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_TEAM_CONVERTED, team.Id, "", "", nil)
+	messageWs.Add("team_id", team.Id)
+	a.Publish(messageWs)
+
+	return team, nil
+}
```
Comment (saturninoabril) [inline app/team.go]: This works fine but the internal implementation is for general purpose of patching any field of the team.  I'd prefer to narrow down its scope.  I think it's best to do it similar to `UpdateTeamScheme`.

File: cmd/mattermost/commands/team.go [hunk 4]
```diff
@@ -339,3 +353,43 @@ func archiveTeamCmdF(command *cobra.Command, args []string) error {
...
+		return errors.New("You can only change the type of public/private teams.")
+	}
+
+	team.Type = model.TEAM_OPEN
+	if private {
+		team.Type = model.TEAM_PRIVATE
+	}
```
Comment (saturninoabril) [inline cmd/mattermost/commands/team.go]: Should we be more explicit here? Like:
```go
if private {
    team.Type = model.TEAM_PRIVATE
} else if public {
    team.Type = model.TEAM_OPEN
} 
```

File: cmd/mattermost/commands/team.go [hunk 5]
```diff
@@ -339,3 +353,43 @@ func archiveTeamCmdF(command *cobra.Command, args []string) error {
...
+
+	user := getUserFromUserArg(a, username)
```
Comment (saturninoabril) [inline cmd/mattermost/commands/team.go]: I think we need to move this up next to username validation and have the user validated as well. Exit early if user == nil.

**Comment by waseem18** [comment]
I see that the rest API to update a team doesn't allow to update the type of a team. 

This API uses the `UpdateTeam` method from `app/team.go` and I was trying to use the same method in the `team modify` command after adding below change to the method 
```
oldTeam.Type = team.Type
```
and this caused the `TestUpdateTeam` from `api4/team_test.go` to fail.

I think we should have a different `Update` methods for rest api and command line or should allow team type to be modified via rest api.

Any thoughts @saturninoabril @grundleborg ?

**Comment by saturninoabril** [comment]
Don't use the `UpdateTeam` method from `app/team.go`.  Instead, use the one from the store.
```go
a.Srv.Store.Team().Update(oldTeam)
```

Have you tried similar implementation of `UpdateTeamScheme`?

**Comment by waseem18** [comment]
I haven't yet tried `UpdateTeamScheme` method though.

I'll use the store's `Update` method

**Comment by waseem18** [comment]
Sure @grundleborg 

I also had one more question, In the command `team modify myteam --private --username myusernamedev` what would be the use of specifying a username?

I the `channel modify` command we use the username to create a post in the respective channel saying that the channel has been modified by so and so username.

How do we deal with having `--username` flag on `team modify` command.

Thanks

**Comment by jasonblais** [comment]
@saturninoabril & @grundleborg would you have thoughts on the above question from Wasim?

**Comment by grundleborg** [comment]
@jasonblais see my above comment, we need to hold this PR for now as the definition of public/private teams is problematic.

**Comment by hanzei** [comment]
@grundleborg Are there updates on the public/private teams definition?

**Comment by grundleborg** [comment]
@hanzei @mkraft is working on sorting out the public/private teams distinction, and once that's done we'll come back to this PR.

**Comment by mkraft** [comment]
We're meeting about it today so I'll give an update here too.

**Comment by hanzei** [comment]
Hey @mkraft,

What was the result of the meeting? What changes should be done for this PR?

**Comment by mkraft** [comment]
@hanzei @waseem18 There is proposal to eliminate `(*Team).Type` which is currently being assed from a product management perspective by @wiersgallak.

**Comment by joshuabezaleel** [comment]
Hi everyone, is there any update on this issue? I would like to give a shot on this if it's possible.

**Comment by jespino** [comment]
Hi, I've been working in the refactor needed to remove the `team.Type` field and make it more "simple" using a `IsPublic` boolean field. But this change, because the API implications has been deferred until the next major (breaking changes) release. So, my proposal here would be. The `--private` would set the team attributes `AllowOpenInvite = false` and the `Type = model.TEAM_INVITE`, and the`--public` would set the team attributes `AllowOpenInvite = true` and the `Type = model.TEAM_OPEN`.

**Comment by grundleborg** [comment]
I'm closing this PR as we have a separate ticket to track the fixing of the private/public handling of teams for Mattermost v6.0 whenever that happens, as it will be a backwards-incompatible change.

**Pull request closed without merging**